spinalcordtoolbox/scripts/sct_deepseg.py

```diff
@@ -130,6 +132,65 @@ def get_parser():
     return parser
 
 
+def segment_and_average_volumes(model_paths, input_filenames, options):
+    """
+        Run `ivadomed.inference.segment_volume()` once per model, then average the outputs.
+
+        :param model_paths: A list of folder paths. The folders must contain:
+            (1) the model ('folder_model/folder_model.pt') to use
+            (2) its configuration file ('folder_model/folder_model.json') used for the training,
+            see https://github.com/neuropoly/ivadomed/wiki/configuration-file
+        :param input_filenames: list of image filenames (e.g. .nii.gz) to segment. Multichannel models require multiple
+            images to segment, i.e.., len(fname_images) > 1.
+        :param options: A dictionary containing optional configuration settings, as specified by the
+            ivadomed.inference.segment_volume function.
+
+        :return: list, list: List of nibabel objects containing the soft segmentation(s), one per prediction class, \
+            List of target suffix associated with each prediction
+    """
+    # Fetch the name of the model (to be used in logging)
+    name_model = Path(model_paths[0]).parts[-1]
+    logger.info(f"\nRunning inference for model '{name_model}'...")
+
+    # Perform inference once per model
+    nii_lsts, target_lsts = [], []
+    for path_model in model_paths:
+        if len(model_paths) > 1:  # We have an ensemble, so output messages to distinguish between seeds
+            name_seed = Path(path_model).parts[-2]
+            logger.info(f"\nUsing '{name_seed}'...")
+        nii_lst, target_lst = imed_inference.segment_volume(path_model, input_filenames, options=options)
+        nii_lsts.append(nii_lst)
+        target_lsts.append(target_lst)
+
+    # If we have a single model, skip averaging
+    if len(model_paths) == 1:
+        nii_lst = nii_lsts[0]
+    # Otherwise, we have a model ensemble, so average the image data
+    else:
+        logger.info(f"\nAveraging outputs across the ensemble for '{name_model}'...")
+        nii_lst = []
+        # NB: `nii_lsts` is a list of lists, with each sublist being the *per-model* predictions. Example:
+        #         [
+        #             [m1_prediction_1.nii.gz, m1_prediction_2.nii.gz, ...],  # model 1 predictions
+        #             [m2_prediction_1.nii.gz, m2_prediction_2.nii.gz, ...]   # model 2 predictions
+        #         ]
+        # So, we want to take: the average of "prediction_1", the average of "prediction_2", etc.
+        # To do this, we unpack + zip `nii_lists`, so that "prediction_N" files are grouped as "predictions".
+        for predictions in zip(*nii_lsts):
+            # Average the data for each output in the ensemble
+            data_stack = np.stack([pred.get_fdata() for pred in predictions], axis=-1)
+            data_mean = np.mean(data_stack, axis=-1)
```

Comment:
Is there a reason you're picking `axis=-1` here, instead of the default `axis=0`? Anecdotally, in a micro-benchmark I just ran on my laptop, `np.stack(axis=-1)` is 3-4 times slower than `np.stack(axis=0)`; and the combination of `np.stack` and `np.mean` is about 2 times slower with `axis=-1` than `axis=0`.

(Why did I try this? My educated guess is that `np.stack(axis=0)` can copy each input array as a single contiguous chunk of memory inside the output array, but that `axis=-1` means that the input arrays are interleaved inside the output array.)

Reply:
There was a reason? But, the reason is pretty shallow, hehe:

4D MRI images use the shape `[x, y, z, t]`, where the `t` dimension distinguishes between 3D `[x, y, z]` volumes across different points in time. So, by convention, I'm used to "stacks of 3D volumes" being represented as stacked along `axis=-1`.

I'm very happy to change this to `axis=0` for performance reasons. Thanks for catching this! :hearts:

Reply:
Addressed by [07a959b](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4042/commits/07a959b9a9d72dba9d582227d6663541b9e1635a).